### DISCUSSION TITLE
Missing error handling/data integrity

### DISCUSSION THREAD

**Discussion by hdk-pd** [spatie/laravel-medialibrary#3010] [Q&A] [answered]
First: Thanks for all the open source work you guys at Spatie do. Much appreciated.

I think there is an issue with error handling/data integrity on media creation. In case I'm not missing a main point of the process/library, there is no error handling when it comes down to the actual upload to the final filesystem.

Laravel: 9.22.1
Laravel MediaLibrary: 10.4.3
Filesystem: S3

A simplified part of my code:
```php
$estate->addMedia($file)->usingFileName($this->generateFileName($file))->toMediaCollection($attachmentType->collectionName(), $attachmentType->disk());
```

File is an instance of `\Symfony\Component\HttpFoundation\File\UploadedFile`, uploaded with FilePond in the frontend. No matter what issue I have with persisting the file on S3 (being it me changing the API Key in my .env file, altering my permissions in AWS IAM or something similar, which obviously causes the file not to be saved on AWS), does not trigger any different functionality than a successful upload. The result of above command is a `\Spatie\MediaLibrary\MediaCollections\Models\Media` instance and the database entry is being kept. Same when the upload succeeds, of course.
Again, I'm unsure what point I might be missing here, but I think there should at least be some kind of data integrity right at the creation of the media. We obviously can't possibly know about S3 losing our file afterwards, but this is very unhelpful. In my mind the creation should be part of a transaction, an exception should be thrown when the upload to the filesystem fails, the media entry rollbacked and the exception thrown again, just like the library throws exceptions such as `\Spatie\MediaLibrary\MediaCollections\Exceptions\FileIsTooBig`.

[`toMediaCollection()`](https://github.com/spatie/laravel-medialibrary/blob/main/src/MediaCollections/FileAdder.php#L285) creates the media model and calls [`attachMedia()`](https://github.com/spatie/laravel-medialibrary/blob/main/src/MediaCollections/FileAdder.php#L409) and basically ends up in [`processMediaItem()`](https://github.com/spatie/laravel-medialibrary/blob/main/src/MediaCollections/FileAdder.php#L428).
That method actually throws a few exceptions in subcalls such as `guardAgainstDisallowedFileAdditions()`. But again, ultimatively, it just persists the model afterwards in line 438 and then calls `add()` on the filesystem in the next lines, which uploads the file to S3. That again comes down to calling `put()` on the Illuminate Filesystem contract. According to that contract, the method should return a boolean. [The return value of that method is never being checked.](https://github.com/spatie/laravel-medialibrary/blob/main/src/MediaCollections/Filesystem.php#L143) This appears to be the main issue I am describing. A `dd()` of the return value prints `false`. So the exceptions of the S3 implementation that are caught under the hood and cause the method to return false are being thrown just as expected, but are being ignored by the library.

I am sorry if I am missing something here and would be glad to be educated on how to handle upload errors (without checking the file existence with the S3 API manually after creation).

**Comment by Silver343** [comment, marked as answer]
@hdk-pd I had a similar issue so made a pr #3067 can you please see if it solves yours too.

**Reply by hdk-pd** [reply]
Sorry for the late reply, I was addressing other parts of my application in the meantime.
It seems that the issue I described did in fact exist as the request has been merged in the meantime. Thank you for your work on this!
